api/hasura/actions/admin_update_user.ts

```diff
@@ -0,0 +1,175 @@
+import assert from 'assert';
+
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+import { z } from 'zod';
+
+import { authCircleAdminMiddleware } from '../../../api-lib/circleAdmin';
+import { gql } from '../../../api-lib/Gql';
+import {
+  adminUpdateUserSchemaInput,
+  composeHasuraActionRequestBody,
+} from '../../../src/lib/zod';
+
+async function handler(request: VercelRequest, response: VercelResponse) {
+  try {
+    const {
+      input: { object: input },
+    } = composeHasuraActionRequestBody(adminUpdateUserSchemaInput).parse(
+      request.body
+    );
+
+    // Validate no epoches are active for the requested user
+    const { circle_id, address } = input;
+
+    const {
+      users: [user],
+    } = await gql.q('query')({
+      users: [
+        {
+          limit: 1,
+          where: {
+            address: { _ilike: address },
+            circle_id: { _eq: circle_id },
+            // ignore soft_deleted users
+            deleted_at: { _is_null: true },
+          },
+        },
+        {
+          id: true,
+          fixed_non_receiver: true,
+          non_receiver: true,
+          starting_tokens: true,
+          give_token_remaining: true,
+          pending_received_gifts: [
+            {},
+            {
+              epoch_id: true,
+              sender_id: true,
+              sender_address: true,
+              tokens: true,
+            },
+          ],
+          circle: {
+            epochs: [
+              {
+                where: {
+                  _and: [
+                    { end_date: { _gt: 'now()' } },
+                    { start_date: { _lt: 'now()' } },
+                  ],
+                },
+              },
+              {
+                start_date: true,
+                end_date: true,
+                id: true,
+              },
+            ],
+          },
+        },
+      ],
+    });
```

Comment:
I can imagine this being reused in other places so probably worth extracting into a separate `queries/getUserInCircle.ts` file